### PR TITLE
DOXYGEN: Add doxygen groups to header files in the common folder

### PR DESCRIPTION
**Pull request by b-gent**
All the header files from the ``common`` folder now have doxygen groups added to them. Thanks to that, the usability of API doc is greatly improved and crucial information can be found. Compare:
Old:
image
New:
image

### CODE REVIEW

**Comment by b-gent** [comment]
@sev-

**Comment by criezy** [comment]
> 1. I couldn't find the doxygen config file used for your current API doc anywhere in the repository 

You can find the current doxygen config file in https://github.com/scummvm/scummvm-sites/tree/doxygen

And indeed there is a lot of work needed to organise  our doxygen documentation and make it usable. This looks like a good start. Using groups is a good idea and the main page could for example be improved with some basic information and link to https://wiki.scummvm.org/index.php?title=Developer_Central.

**Comment by Mataniko** [comment]
Thanks for the PR, Looking forward to seeing your application!

**Comment by b-gent** [comment]
> > 1. I couldn't find the doxygen config file used for your current API doc anywhere in the repository
> 
> You can find the current doxygen config file in https://github.com/scummvm/scummvm-sites/tree/doxygen

I wasn't aware of the scummvm-sites fork, thanks :)

> And indeed there is a lot of work needed to organise our doxygen documentation and make it usable. This looks like a good start. Using groups is a good idea and the main page could for example be improved with some basic information and link to https://wiki.scummvm.org/index.php?title=Developer_Central.

Yes, and not only the main page, I bet there is much more 'conceptual' documentation that could be added to the doxygen docset that would explain how the API works. The empty placeholder pages that are there right now that are supposed to explain  how to implenent a plugin are an example of this.

File: common/array.h [hunk 1]
```diff
@@ -30,6 +30,16 @@
 
 namespace Common {
 
+/**
+ * @defgroup common_array Arrays
+ * @ingroup common
+ *
+ * @brief  Functions for working on arrays.
```
Comment (Mataniko) [inline common/array.h:37]: these are functions to replace some std arrays

File: common/md5.h [hunk 2]
```diff
@@ -27,6 +27,15 @@
 
 namespace Common {
 
+/**
+ * @defgroup common_md5 MD5 checksum
+ * @ingroup common
+ *
+ * @brief API for computing the MD5 checksum.
```
Comment (Mataniko) [inline common/md5.h:34]: This is less of an API and more of a method/class - @sev- WDYT on terminology in general

File: common/memory.h [hunk 3]
```diff
@@ -27,6 +27,14 @@
 
 namespace Common {
 
+/**
+ * @defgroup common_memory Memory
+ * @ingroup common
+ *
+ * @brief Functions for managing the memory.
```
Comment (Mataniko) [inline common/memory.h:34]: Same as earlier - we should have consistent terminology - functions, api, class, methods

**Comment by Mataniko** [comment]
Overall looks like a massive improvement - left a couple comments on terminology that we should standardize before proceeding

File: common/md5.h (hunk 2 above)
Comment (b-gent) [reply common/md5.h:34]: That's a good point about having consistent terminology. As I see it, 'API' is a superset of whatever we might have in these header files - classes, methods, functions, structures etc. If you agree with this, then probably using the term 'API' is safest and most future-proof.

File: common/memory.h (hunk 3 above)
Comment (b-gent) [reply common/memory.h:34]: I guess in this case the most accurate description is 'Class templates for managing the memory' ?

File: common/array.h (hunk 1 above)
Comment (b-gent) [reply common/array.h:37]: I'll correct this (as soon as we agree on terminology).

**Comment by b-gent** [comment]
I would like to get this in as soon as possible. I'm already making additional changes to these files on another branch and I would like to avoid some crazy merge conflicts.

**Comment by b-gent** [comment]
The output is now available for preview on https://b-gent.github.io/PR2361/group__common.html

@sev , it would be great if you could have a look at the comments here and the PR as a whole.

File: common/array.h (hunk 1 above)
Comment (criezy) [reply common/array.h:37]: Actually these are classes and not functions.

File: common/md5.h (hunk 2 above)
Comment (criezy) [reply common/md5.h:34]: In this specific case I would probably have used `Functions` if I had written the documentation. But API is also fine with me. And indeed we need consistency.

File: common/memory.h (hunk 3 above)
Comment (criezy) [reply common/memory.h:34]: Referring to classes here seems confusing to me. They look like `Template functions`.

**Comment by sev-** [comment]
Great work, merging

**Pull request merged**